### PR TITLE
8252696: Loop unswitching may cause out of bound array load to be executed

### PR DESCRIPTION
**Pull request by rwestrel**
Review thread so far: https://mail.openjdk.java.net/pipermail/hotspot-compiler-dev/2020-September/039853.html

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8252696](https://bugs.openjdk.java.net/browse/JDK-8252696): Loop unswitching may cause out of bound array load to be executed

### Reviewers
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/176/head:pull/176`
`$ git checkout pull/176`

### CODE REVIEW

File: src/hotspot/share/opto/loopPredicate.cpp [hunk 1]
```diff
@@ -249,7 +250,7 @@ void PhaseIdealLoop::clone_concrete_loop_predicates(Deoptimization::DeoptReason
     uncommon_proj = iff->proj_out(1 - predicate->as_Proj()->_con);
     if (uncommon_proj->unique_ctrl_out() != rgn)
       break;
-    if (iff->is_RangeCheck()) {
+    if (iff->in(1)->Opcode() == Op_Opaque4 && skeleton_predicate_has_opaque(iff)) {
       // Only need to clone range check predicates as those can be changed and duplicated by inserting pre/main/post loops
```
Comment (neliasso) [inline src/hotspot/share/opto/loopPredicate.cpp:254]: Is the comment still correct? It is still always a range check predicate?

Comment (rwestrel) [reply src/hotspot/share/opto/loopPredicate.cpp:254]: Hi Nils,
Yes, the comment is still correct. Skeleton predicates are always range checks. But in some cases to protect against overflow, predicates perform range checks with long computations in which case they use an IfNode and not a RangeCheckNode.

**Review by neliasso — APPROVED** [review]
Thanks for your answer! 

Looks good!

**Review by chhagedorn — APPROVED** [review]
That looks good to me!

As we have discussed offline, we should probably do the additional clean-ups of this code in a separate RFE.

**Comment by rwestrel** [comment]
@chhagedorn @neliasso Thanks for the reviews

**Pull request closed without merging**